Use appropriate dask instance based on its computation result / Improve tests

sinhrks opened the pull request:
Follow up for #506 and #509. I've misunderstood that ``eq`` can't be used for ``DataFrame`` tests...

## Summary

I've included following fixes to this PR because these are related each other. I'm willing to split if required .
- [x] Made appropriate dask classes to be used to wrap pandas objects, ``dd.DataFrame`` tfor ``pd.DataFrame`` and ``dd.Series`` for ``pd.Series``.
- [x] ``eq`` in ``test_dataframe.py`` now validates ``dask`` instance and its computation results. Fixed some bugs detected by this additions (below).
- [x] change scalar.__init__ to be compat with Series an DataFrame
- [x] ``DataFrame`` and ``Series`` now has ``constructor_sliced`` ``constructor`` properties (names are derived from pandas)
- [x] ``map_partition`` and ``apply_concat_apply`` can now accept optional ``return_type`` arg to wrap computation result with *correct* ``dask`` instance.
- [x] Added some tests to be comprehensive a bit.
- [x] Bug: ``DataFrame.loc`` with scalar, like ``df.loc[1]`` results in ``DataFrame``. Fixed to return ``Series``
- [x] Bug: ``SeriesGroupBy`` aggregations result in ``DataFrame``. Fixed to return ``Series``.

mrocklin commented on dask/dataframe/tests/test_dataframe.py:
```diff
@@ -69,7 +69,7 @@ def test_Series():
     assert isinstance(d.a + 1, dd.Series)
 
-    tm.assert_frame_equal((d + 1).compute(), full + 1)
+    eq((d + 1), full + 1)
```
Missing `assert`

mrocklin commented:
It seems ok to me to use `tm.assert_frame_equal`.  We should also consider improving `eq` or replacing it with our own `assert_frame_equal` function for dask.array.

sinhrks commented:
Yep. One idea from #513 is checking the consistency of ``dask`` instance and its computation result. For example, currently ``dd.DataFrame.loc[1]`` results in ``dd.DataFrame``, but the computation result is ``pd.Series``.

I've added a validation logic and fixed:

- ``DataFrame.loc[1]`` now return ``Series`` rather than ``DataFrame``, by adding some [constructor property same as pandas](http://pandas.pydata.org/pandas-docs/stable/internals.html#override-constructor-properties)
- ``SeriesGroupBy`` aggregations now return ``Series`` rather than ``DataFrame``, by adding ``return_type`` kw to ``apply_concat_apply``.

If it looks reasonable, I'll check other part and add more tests.

mrocklin commented:
This all looks reasonable to me.

mrocklin commented:
A couple of large PRs were recently merged which have likely caused merge conflicts.  I apologize for the inconvenience.  I'm happy to help with merge/rebase if necessary.

sinhrks commented:
Thanks, rebased. I've summarized the changes in PR on the top.

mrocklin commented:
This looks great to me.  Merging soon if no comments.
